File: 236-lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        //return helper(root,p,q);

        return helper2(root, p, q);
    }

    TreeNode helper(TreeNode root, TreeNode p, TreeNode q){


        if (root == null ) return null;
        if (root ==p || root==q){
            return root;
        }

        if (root.left ==null && root.right == null) return null;



        TreeNode left = helper(root.left,p,q);
        TreeNode right = helper(root.right,p,q);

        if (left!=null && right!=null){
            return root;

        }
        if (left!=null) return left;
        if (right!=null) return right;

        return null;

    }

    TreeNode helper2(TreeNode root, TreeNode p, TreeNode q){

        if (root == null) return root;
        if (root == p || root == q) return root;

       TreeNode l = helper2(root.left, p, q);
       TreeNode r = helper2(root.right, p, q);

       if (l != null && r != null ) return root;
      
           if (l != null ) return l;
           else if (r != null) return r;
       return null;
    }
}

File: 236-lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        

        return self.helper(root, p, q)

    def helper(self, root, p, q):

        if root is None:
            return None

        if root  == p or root == q:
            return root

        
        l = self.helper(root.left, p, q)
        r = self.helper(root.right, p, q)

        if l and r:
            return root

        if l:
            return l
        if r:
            return r

        



